src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -802,6 +829,40 @@ public void Close()
                 _terminateDelegate(_nodeId);
             }
 
+            /// <summary>
+            /// Waits for the child node process to exit.
+            /// </summary>
+            public async Task WaitForExitAsync()
+            {
+                // Wait for the process to exit.
+                if (_isExiting)
+                {
+                    CommunicationsUtilities.Trace("Waiting for node with pid = {0} to exit", _process.Id);
+
+                    // .NET 5 introduces a real WaitForExitAsyc.
+                    // This is a poor man's implementation that uses polling.
+                    int timeout = TimeoutForWaitForExit;
+                    int delay = 5;
+                    while (timeout > 0)
+                    {
+                        bool exited = _process.WaitForExit(milliseconds: 0);
+                        if (exited)
+                        {
+                            return;
+                        }
+                        timeout -= delay;
+                        await Task.Delay(delay).ConfigureAwait(false);
+
+                        // Double delay up to 500ms.
+                        delay = Math.Min(delay * 2, 500);
+                    }
+                }
+
+                // Kill the child and do a blocking wait.
+                CommunicationsUtilities.Trace("Killing node with pid = {0}", _process.Id);
```

Comment:
@Forgind I think this takes care of https://github.com/dotnet/msbuild/pull/6053#issuecomment-771762934?

Reply:
@cdmihai commented (https://github.com/dotnet/msbuild/pull/6053/files#r566958454) we may want to do something with higher visibility?

Reply:
I talked with @cdmihai about this. We think it would be good to have both the comm trace (that you already added) and a warning to make it more visible. NodeProviderOutOfProcBase has an IBuildComponentHost (_componentHost) with access to the LoggingService. You can leave BuildEventContext and BuildEventFileInfo undefined as in the [build manager](https://github.com/dotnet/msbuild/blob/4b9f6122dfc8397ee0d939d7932df9cd5818c90c/src/Build/BackEnd/BuildManager/BuildManager.cs#L2855-L2861).